entity-framework/core/modeling/data-seeding.md

```diff
@@ -28,3 +28,32 @@ To add entities that have a relationship the foreign key values need to be speci
 Once entities have been added, it is recommended to use [migrations](xref:core/managing-schemas/migrations/index) to apply changes. 
 
 Alternatively, you can use `context.Database.EnsureCreated()` to create a new database containing the seed data, for example for a test database or when using the in-memory provider. Note that if the database already exists, `EnsureCreated()` will neither update the schema nor the seed data in the database.
+
+This type of seed data is managed by migrations and the script to update the data that's already in the database needs to be generated without connecting to the database. This imposes some restrictions:
+* The primary key value needs to be specified even if it's usually generated by the database
+* Any changes to the seed data done outside of migrations will be lost if the primary key is changed in any way
+
+Therefore this feature is most useful for static data that's not expected to change outside of migrations and doesn't depend on anything else in the database.
+
+If you need to seed temporary data for testing, complex data that depends on database state or data that uses alternate keys as identity it is recommended to add the seeding logic to the application start up method.
```

Comment:
Why turn this into a recommendation? It seems there are some of the same potentially concurrency issues as in applying migrations at startup.

If you need to setup seed data for your application and for some reason HasData didn't work for you, you can do it in code in your application, but it doesn't need to be at startup. It can be in some controller action available to the admin of the application instead. 

If you need to setup sample data for a test, you can do it as part of the test's setup.

Reply:
This was the smallest sample I could think of to illustrate the workaround. Do you think we should provide a sample controller action instead?

Reply:
Possibly. But let's discuss the whole issue with @ajcvickers. I have the impression I am missing something.

Reply:
We should mention that doing migrations and or seeding is not recommended in a running application as it could cause concurrency and permission issues.
Ideally this sort of operations should be performed during deployment in some way. And migrations seed data helps doing it is some scenarios.